yt_dlp/extractor/niconicochannelplus.py

```diff
@@ -0,0 +1,455 @@
+import functools
+import json
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    OnDemandPagedList,
+    UserNotLive,
+    filter_dict,
+    int_or_none,
+    parse_iso8601,
+    parse_qs,
+    traverse_obj,
+)
+
+
+class NiconicoChannelPlusBaseIE(InfoExtractor):
+    _WEBPAGE_BASE_URL = 'https://nicochannel.jp'
+
+    def _call_api(self, path, item_id, *args, **kwargs):
+        return self._download_json(
+            f'https://nfc-api.nicochannel.jp/fc/{path}', video_id=item_id, *args, **kwargs)
+
+    def _find_fanclub_site_id(self, channel_name):
+        fanclub_list_json = self._call_api(
+            'content_providers/channels', item_id=f'channels/{channel_name}',
+            note='Fetching channel list', errnote='Unable to fetch channel list',
+        )['data']['content_providers']
+
+        for fanclub in fanclub_list_json:
+            if fanclub.get('domain') == f'{self._WEBPAGE_BASE_URL}/{channel_name}':
+                return fanclub['id']
+
+        raise ExtractorError(f'Non-existing channel: {channel_name}', expected=True)
```

Comment:
```suggestion
        fanclub_list_json = self._call_api(
            'content_providers/channels', item_id=f'channels/{channel_name}',
            note='Fetching channel list', errnote='Unable to fetch channel list',
        )
        fanclub_id = traverse_obj(fanclub_list_json, (
            'data', 'content_providers', lambda _, v: v['domain'] == f'{self._WEBPAGE_BASE_URL}/{channel_name}', 'id'),
            get_all=False)
        if not fanclub_id:
            raise ExtractorError(f'Channel {channel_name} does not exist', expected=True)
        return fanclub_id
```

Reply:
Thanks for your awesome example about "`traverse_obj()`".

After thinking about it, I think the following code might be better:
(highlight works not well)

```suggestion
        fanclub_list_json = self._call_api(
            'content_providers/channels', item_id=f'channels/{channel_name}',
            note='Fetching channel list', errnote='Unable to fetch channel list',
        )['data']['content_providers']
        fanclub_id = traverse_obj(fanclub_list_json, (
            lambda _, v: v.get('domain') == f'{self._WEBPAGE_BASE_URL}/{channel_name}', 'id'),
            get_all=False)
        if not fanclub_id:
            raise ExtractorError(f'Channel {channel_name} does not exist', expected=True)
        return fanclub_id
```

Why:

- It should be fatal if "```json['data']['content_providers']```" does not exist, because we need the data in it.
- It should not be fatal if "```json['data']['content_providers'][\d]['domain']```" does not exist, because we are judging equality and there's no reason to be held back by false items from the list.

Your opinion?

Reply:
with Lesmiscore's suggestion, it would still be fatal if `fanclub_list_json['data']['content_providers']` doesn't exist, because the only data we are extracting is the `fanclub_id`, and we will raise an ExtractorError if traverse_obj returns None. 

and if the ExtractorError is to be expected, then it would be better not to risk a KeyError beforehand

Reply:
Hi, @bashonly , thanks for your reply.

> with Lesmiscore's suggestion, it would still be fatal if `fanclub_list_json['data']['content_providers']` doesn't exist, because the only data we are extracting is the `fanclub_id`, and we will raise an ExtractorError if traverse_obj returns None.

Yes.

> `lambda _, v: v['domain'] == f'{self._WEBPAGE_BASE_URL}/{channel_name}'` is non-fatal

Yes, it is.

> And if the ExtractorError is to be expected, then it would be better not to risk a KeyError beforehand.

Hmm, it depends.

If the "```fanclub_list_json['data']['content_providers']```" object exists but no matching channel is found, we should throw an expected ExtractorError; if that JSON object does not exist, we can suspect that the website (nicochannel.jp) has changed its API.
The "```_find_fanclub_site_id()```" function is the first one who handles user input, and also the first one who fetches data from the website. If the fetched data looks good, users should re-check their input; if the data is unexpected, we will have a chance to receive bug reports.

Here are the examples. If there is no "```['data']['content_providers']```" in the returned JSON object, when using the suggestion:

```console
[debug] [NiconicoChannelPlus] Extracting URL: https://nicochannel.jp/foo/live/smart
[NiconicoChannelPlus] channels/foo: Fetching channel list
ERROR: [NiconicoChannelPlus] None: Channel foo does not exist
  ...

```

We got "Channel foo does not exist", but it's not real.

When using my new code:

```console
[debug] [NiconicoChannelPlus] Extracting URL: https://nicochannel.jp/foo/live/smart
[NiconicoChannelPlus] channels/foo: Fetching channel list
ERROR: An extractor error has occurred. (caused by KeyError('data')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  ...
KeyError: 'data'

```

I think the latter one is better.

Your opinion?